pytorch_lightning/trainer/connectors/logger_connector/result.py

```diff
@@ -501,7 +530,37 @@ def __str__(self) -> str:
     def __getstate__(self) -> dict:
         d = self.__dict__.copy()
         # can't deepcopy tensors with grad_fn
-        minimize = d.get('_minimize')
+        minimize = d['_minimize']
         if minimize is not None:
             d['_minimize'] = minimize.detach()
-        return d
+        extra = self.get('_extra')
+        if extra is not None:
+            d['_extra'] = extra
+        # all the items should be either `ResultMetric`s or `ResultMetricCollection`s
+        items = {k: v.__getstate__() for k, v in self.items() if k != '_extra'}
+        return {**d, 'items': items}
+
+    def __setstate__(self, state: dict) -> None:
+        self.__dict__.update({k: v for k, v in state.items() if k != 'items'})
+
+        def setstate(item: dict) -> Union[ResultMetric, ResultMetricCollection]:
+            if not isinstance(item, dict):
+                raise ValueError(f'Unexpected value: {item}')
+            cls = item['_class']
+            if cls == ResultMetric.__name__:
+                result_metric = ResultMetric(item['meta'], item['is_tensor'])
+            elif cls == ResultMetricCollection.__name__:
+                result_metric = ResultMetricCollection()
+            else:
+                raise ValueError(f"Unexpected class name: {cls}")
+            result_metric.__setstate__(item)
+            return result_metric
+
+        items = {k: setstate(v) for k, v in state['items'].items()}
+        self.update(items)
+
```

Comment:
```suggestion
        items = {k: setstate(v) for k, v in state['items'].items()}
        self.update(items)
        
        self.to(self.device)
```

Reply:
Done. Also implemented `map_location` and added a test for it